src/read/xz.rs

```diff
@@ -0,0 +1,330 @@
+}
+
+impl<R: Read> XzDecoder<R> {
+    pub fn new(inner: R) -> Self {
+        XzDecoder {
+            compressed_reader: BufReader::new(inner),
+            stream_size: 0,
+            buf: VecDeque::new(),
+            check_size: 0,
+            records: vec![],
+            flags: [0, 0],
+        }
+    }
+}
+
+struct CountReader<'a, R: BufRead> {
+    inner: &'a mut R,
+    count: &'a mut usize,
+}
+
+impl<R: BufRead> Read for CountReader<'_, R> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let count = self.inner.read(buf)?;
+        *self.count += count;
+        Ok(count)
+    }
+}
+
+impl<R: BufRead> BufRead for CountReader<'_, R> {
+    fn fill_buf(&mut self) -> Result<&[u8]> {
+        self.inner.fill_buf()
+    }
+
+    fn consume(&mut self, amt: usize) {
+        self.inner.consume(amt);
+        *self.count += amt;
+    }
+}
+
+struct BufWriter<'a> {
+    inner: &'a mut [u8],
+    written: &'a mut usize,
+    total: &'a mut usize,
+    rest: &'a mut VecDeque<u8>,
+}
+
+impl<'a> Write for BufWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        if self.inner.len() > *self.written {
+            let len = std::cmp::min(buf.len(), self.inner.len() - *self.written);
+            self.inner[*self.written..*self.written + len].copy_from_slice(&buf[..len]);
+            *self.written += len;
+            *self.total += buf.len();
+            Ok(len)
+        } else {
+            self.rest.extend(buf.iter().copied());
+            *self.total += buf.len();
+            Ok(buf.len())
+        }
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        Ok(())
+    }
+}
+
+fn get_multibyte<R: Read>(input: &mut R, mut hasher: Option<&mut Hasher>) -> Result<u64> {
+    let mut result = 0;
+    for i in 0..9 {
+        let mut b = [0u8; 1];
+        input.read_exact(&mut b)?;
+        if let Some(hasher) = hasher.as_mut() {
+            hasher.update(&b);
+        }
+        let b = b[0];
+        result ^= ((b & 0x7F) as u64) << (i * 7);
+        if (b & 0x80) == 0 {
+            return Ok(result);
+        }
+    }
+    Err(Error::new(
+        std::io::ErrorKind::InvalidData,
+        "Invalid multi-byte encoding",
+    ))
+}
+
+impl<R: Read> Read for XzDecoder<R> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        if !self.buf.is_empty() {
+            let len = std::cmp::min(buf.len(), self.buf.len());
+            for (i, b) in self.buf.drain(..len).enumerate() {
+                buf[i] = b;
+            }
+            return Ok(len);
+        }
+        let mut reader = CountReader {
+            inner: &mut self.compressed_reader,
+            count: &mut self.stream_size,
+        };
+        if *reader.count == 0 {
+            let mut b = [0u8; 6];
+            match reader.read(&mut b) {
+                Ok(0) => return Ok(0),
+                Err(e) => return Err(e),
+                _ => (),
+            }
+            if b != [0xfd, b'7', b'z', b'X', b'Z', 0] {
```

Comment:
Combine the magic bytes, flags and CRC32 into a 12-byte read, for performance reasons. You can use the `to_and_from_le!` macro to help with this. Also, please join the byte literals into a byte-slice literal: `b"\xFD7zXZ\0"`